Health filter for EBS

taohungyang opened the pull request:
Continue https://github.com/capitalone/cloud-custodian/pull/911#discussion_r100299694

- Register filter "health-event" in `resources/ebs.py`

- Inherit `HealthEventFilter` from `filters/health.py` and modify `def process`:

  - Not using the resource id filters since the existed ebs resources might not contains the affected resources (eg: AWS_EBS_VOLUME_LOST)
  - Load the resource description from config
- Add validation to avoid event type that does not belong to the resource

Also some minor bugs fixed.
Please let me know any comments or suggestions. Thanks @kapilt

kapilt commented on c7n/filters/health.py:
```diff
@@ -29,12 +29,20 @@ class HealthEventFilter(Filter):
                    'health:DescribeEventDetails')
 
+    def validate(self):
+        if self.data.get('types'):
```
this seems suspect. what are the values for types in the config here? this could use a doc string. its also unclear what the purpose is. effectively it seems to be saying .. resource_type in types but types here corresponds to event types.

taohungyang replied:
The intention of this is to avoid cases like the following:
![image](https://cloud.githubusercontent.com/assets/10120438/23173650/e105d6b2-f827-11e6-876b-9bc76c30cc38.png)
Which apparently the event type doesn't belong to the resource_type/service.

The format for all the event types is like: AWS_{service}_{Event Code}
For example:
AWS_EBS_VOLUME_LOST
AWS_EC2_PERSISTENT_INSTANCE_RETIREMENT_SCHEDULED
AWS_RDS_MAINTENANCE_SCHEDULED

So I though it might be a possible way to validate by checking if the resource_type/service is part of the event type.

kapilt replied:
the issue is that is no guarantee of correspondence from the resource type name in custodian to the health event type that a user might be interested in.

kapilt replied:
in the context of ebs, it might be better to just put in the valid resource events as an enum in the schema.

kapilt commented on c7n/resources/ebs.py:
```diff
@@ -446,6 +449,57 @@ def process(self, resources, event=None):
+        config = local_session(self.manager.session_factory).client('config')
+        paginator = config.get_paginator('get_resource_config_history')
+        for p in paginator.paginate(
+            resourceType='AWS::EC2::Volume',resourceId=rid):
```
re style, we try to follow pep8, useful to add flake8 tool to your editor of choice

taohungyang replied:
will do. Thanks!

kapilt commented on c7n/resources/ebs.py:
```diff
@@ -446,6 +449,57 @@ def process(self, resources, event=None):
+        m = self.manager
+        result_map = {}
+
+        f = {'services': [m.data['resource'].upper()],
```
this implementation seems like a lot of code duplication from the base class, i'd suggest refactoring the base class so you can get the custom logic for lost volume without the duplication.

kapilt commented on c7n/resources/ebs.py:
```diff
@@ -446,6 +449,57 @@ def process(self, resources, event=None):
+    def load_resource(self, rid):
+        config = local_session(self.manager.session_factory).client('config')
+        paginator = config.get_paginator('get_resource_config_history')
+        for p in paginator.paginate(
```
this is loading all history for the resource, when effectively you just want the last revision, don't need a paginator then, just pass limit=1

kapilt replied:
configuration is null in the deleted resource case and we should probably do limit=2 to handle that case, as the lost volume scenario will effectively always have a deleted resource in index 0 when doing default reverse date sort. we should be walking through the event pair retrieved to a non deleted one.

taohungyang commented:
Modified the following parts:

1. Remove the validation code and put valid event type codes of ebs in the schema.
2. Refactor the base class to reduce duplication.
3. Load only the latest two configurations from history and retrieve the non-deleted one.
4. Restyle

Also, fixed the following bug:

1. describe_affected_entities only return 10 entities at most. Use paginator instead to get the whole list of affected entities.

Thanks @kapilt

kapilt commented on tests/test_ebs.py:
```diff
@@ -280,3 +280,17 @@ def test_ebs_metrics_percent_filter(self):
+                'types': ['AWS_EBS_VOLUME_LOST']}]
+                },session_factory=session_factory)
+        resources = policy.run()
+        self.assertEqual(len(resources), 0)
```
we probably want to do record this against a place where it actually exhibits and then manually cleanup the recorded files to get a positive assertion test.

taohungyang commented:
Finished recording against a place and cleaning up the records.
Also updated the test case in EBS to cover more.

Thanks, @kapilt

kapilt commented:
Looks good, thanks!

kapilt commented on c7n/filters/health.py:
```diff
@@ -40,39 +41,51 @@ def process(self, resources, event=None):
-                    found.add(rid)
-                seen.update(event_map.keys())
-        return [resource_map[rid] for rid in found]
+    def process_event(self, health_events, entities):
```
process event should just return the entities instead of mutating a parameter, afaics all the callers just pass an empty list.

kapilt commented on c7n/filters/health.py:
```diff
@@ -40,39 +41,51 @@ def process(self, resources, event=None):
-                        'eventDescription']['latestDescription']
-                entities = client.describe_affected_entities(
-                    filter={'eventArns': event_map.keys()})['entities']
+    def get_filter(self):
```
i'd call this get_filter_parameters() to disambiguate
